jaxpm/pm.py

```diff
@@ -1,50 +1,84 @@
+from functools import partial
+
 import jax
 import jax.numpy as jnp
 import jax_cosmo as jc
-from jax_cosmo import Cosmology
-
-from jaxpm.growth import growth_factor, growth_rate, dGfa, growth_factor_second, growth_rate_second, dGf2a
-from jaxpm.kernels import PGD_kernel, fftk, gradient_kernel, invlaplace_kernel, longrange_kernel
-from jaxpm.painting import cic_paint, cic_read
-
-
-
-def pm_forces(positions, mesh_shape, delta=None, r_split=0):
+from jax.sharding import PartitionSpec as P
+
+from jaxpm.distributed import (autoshmap, fft3d, get_local_shape, ifft3d,
+                               normal_field)
+from jaxpm.growth import (dGf2a, dGfa, growth_factor, growth_factor_second,
+                          growth_rate, growth_rate_second)
+from jaxpm.kernels import (PGD_kernel, fftk, gradient_kernel,
+                           invlaplace_kernel, longrange_kernel)
+from jaxpm.painting import cic_paint, cic_paint_dx, cic_read, cic_read_dx
+
+
+def pm_forces(positions,
+              mesh_shape=None,
+              delta=None,
+              r_split=0,
+              halo_size=0,
+              sharding=None):
     """
     Computes gravitational forces on particles using a PM scheme
     """
+    if mesh_shape is None:
+        assert (delta is not None),\
+          "If mesh_shape is not provided, delta should be provided"
+        mesh_shape = delta.shape
+
     if delta is None:
-        delta_k = jnp.fft.rfftn(cic_paint(jnp.zeros(mesh_shape), positions))
+        field = cic_paint_dx(positions, halo_size=halo_size, sharding=sharding)
+        delta_k = fft3d(field)
     elif jnp.isrealobj(delta):
-        delta_k = jnp.fft.rfftn(delta)
+        delta_k = fft3d(delta)
     else:
         delta_k = delta
 
+    kvec = fftk(delta_k)
     # Computes gravitational potential
-    kvec = fftk(mesh_shape)
-    pot_k = delta_k * invlaplace_kernel(kvec) * longrange_kernel(kvec, r_split=r_split)
+    pot_k = delta_k * invlaplace_kernel(kvec) * longrange_kernel(
+        kvec, r_split=r_split)
     # Computes gravitational forces
-    return jnp.stack([cic_read(jnp.fft.irfftn(- gradient_kernel(kvec, i) * pot_k), positions) 
-                      for i in range(3)], axis=-1)
+    forces = jnp.stack([
+        cic_read_dx(ifft3d(-gradient_kernel(kvec, i) * pot_k),
+                    halo_size=halo_size,
+                    sharding=sharding) for i in range(3)
+    ],
+                       axis=-1)
+
+    return forces
 
 
-def lpt(cosmo:Cosmology, init_mesh, positions, a, order=1):
+def lpt(cosmo, initial_conditions, a, halo_size=0, sharding=None, order=1):
```

Comment:
Can we do this in a way that preserves the previous API? Maybe with a deprecation warning of positional argument for the next major release of jaxpm?

Reply:
It's also fine that distribution doesn't work with the old api

Reply:
Previous API is changed with a very small change 
make_ode needs to get the original positions in case of a absolute CIC_PAINT 
Distribution works for both CIC implementation (absolute or relative)